parasoft-findings-teamcity-agent/src/main/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcess.java

```diff
@@ -253,6 +263,57 @@ private void reportUnexpectedFormat(File from) {
         _build.getBuildLogger().error("Unexpected report format: "+from.getAbsolutePath()+ ". \nPlease try recreating the report. If this does not resolve the issue, please contact Parasoft support.");
     }
 
+    private void parseAndSendViolationMessages(File pmdReport) {
+        try {
+            Map<String, CodeInspectionType> codeInspectionTypeMap = new HashMap<String, CodeInspectionType>();
+            List<CodeInspectionInstance> codeInspectionInstances = new ArrayList<CodeInspectionInstance>();
+            Document document = getDocument(pmdReport);
+
+            // Handle file elements
+            NodeList fileNodes = getNodes(document, "/pmd/file");
+            for(int i = 0; i < fileNodes.getLength(); i++) {
+                Node fileNode = fileNodes.item(i);
+                if(Node.ELEMENT_NODE != fileNode.getNodeType()) {
+                    continue;
+                }
+                // Handle violation elements
+                NodeList violationNodes = fileNode.getChildNodes();
+                for(int j = 0; j < violationNodes.getLength(); j++) {
+                    Node violationNode = violationNodes.item(j);
+                    if(Node.ELEMENT_NODE != violationNode.getNodeType()) {
+                        continue;
+                    }
+                    NamedNodeMap fileAttributes = fileNode.getAttributes();
+                    NamedNodeMap violationAttributes = violationNode.getAttributes();
+                    Node violationContent = violationNode.getFirstChild();
+                    String cit_rule = violationAttributes.getNamedItem("rule").getNodeValue();
+                    String cit_category = violationAttributes.getNamedItem("ruleset").getNodeValue();
+                    String cit_description = violationAttributes.getNamedItem("ruledescription").getNodeValue();
+                    String ci_message = violationContent.getNodeValue();
+                    String ci_line = violationAttributes.getNamedItem("beginline").getNodeValue();
+                    String ci_fileLocation = fileAttributes.getNamedItem("name").getNodeValue();
+                    String ci_severityNumber = violationAttributes.getNamedItem("priority").getNodeValue();
+
+                    codeInspectionTypeMap.put(cit_rule, new CodeInspectionType(cit_rule, cit_description, cit_category));
```

Comment:
Can we change the map for inspection type to be a set which checks rule ID and send the service message here if id doesn't exist?
And for inspection can we send service message to report it here directly?

Reply:
Yes, it can be.